examples/belief_propagation/bpexample.jl

```diff
@@ -46,23 +47,62 @@ function main()
   vertex_groups = nested_graph_leaf_vertices(
     partition(partition(ψψ, group(v -> v[1], vertices(ψψ))); nvertices_per_partition=nsites)
   )
-  mts = compute_message_tensors(ψψ; vertex_groups=vertex_groups)
-  sz_bp =
-    calculate_contraction(
-      ψψ, mts, [(v, 1)]; verts_tensors=ITensor[apply(op("Sz", s[v]), ψ[v])]
-    )[] / calculate_contraction(ψψ, mts, [(v, 1)])[]
+  mts = belief_propagation(ψψ; vertex_groups=vertex_groups)
+  numerator_network = approx_network_region(
+    ψψ, mts, [(v, 1)]; verts_tn=ITensorNetwork([apply(op("Sz", s[v]), ψ[v])])
+  )
+  denominator_network = approx_network_region(ψψ, mts, [(v, 1)])
+  sz_bp = contract(numerator_network)[] / contract(denominator_network)[]
 
   println(
-    "General Belief Propagation (2-site subgraphs) Gives Sz on Site " *
+    "General Belief Propagation (4-site subgraphs) Gives Sz on Site " *
     string(v) *
     " as " *
     string(sz_bp),
   )
 
-  #Now do it exactly
+  #Now do General Belief Propagation with Matrix Product State Message Tensors Measure Sz on Site v
+  ψψ = flatten_networks(ψ, dag(ψ); combine_linkinds=false, map_bra_linkinds=prime)
   Oψ = copy(ψ)
   Oψ[v] = apply(op("Sz", s[v]), ψ[v])
-  sz_exact = contract_inner(Oψ, ψ) / contract_inner(ψ, ψ)
+  ψOψ = flatten_networks(ψ, dag(Oψ); combine_linkinds=false, map_bra_linkinds=prime)
+
+  combiners = linkinds_combiners(ψψ)
+  ψψ = combine_linkinds(ψψ, combiners)
+  ψOψ = combine_linkinds(ψOψ, combiners)
+
+  vertex_groups = nested_graph_leaf_vertices(partition(ψψ, group(v -> v[1], vertices(ψψ))))
+  maxdim = 8
+
+  mts = belief_propagation(
+    ψψ;
+    vertex_groups=vertex_groups,
+    contract_kwargs=(;
+      alg="density_matrix",
+      output_structure=path_graph_structure,
+      maxdim,
+      contraction_sequence_alg="optimal",
+    ),
+    init_contract_kwargs=(;
+      alg="density_matrix",
+      output_structure=path_graph_structure,
+      cutoff=1e-16,
+      contraction_sequence_alg="optimal",
+    ),
```

Comment:
Why are there two sets of keyword arguments here?

Reply:
One is for the construction of the initial message tensors (where you generally want to contract the rank 1 boundary network down to the desired `output_structure` which specifies the initial structure of the message tensors) and the other is for the message tensor update step where you are going to want to think harder about contracting the resulting network down to the desired structure.

Reply:
Wasn't entirely sure how to handle this part. It comes down to how one does the initialization of the message tensors.

Reply:
Okay yeah I can see the logic behind that. Let me remove that function and have a go at writing some wrappers.

Do you think the default should be MPS or just a single tensor for the message tensors? I would lean towards a single tensor just as that is the original formulation of BP for a tensornetwork.

Reply:
I would vote for defaulting to an MPS since that scales to large partition sizes. Also it has the nice limit that for a rank-1 MPS, the message tensors of generalized BP reduce to a product of message tensors, which is the same as simple BP.